sopel/irc/abstract_backends.py

```diff
@@ -17,39 +23,67 @@ class AbstractIRCBackend(abc.ABC):
     Some methods of this class **MUST** be overridden by a subclass, or the
     backend implementation will not function correctly.
     """
-    def __init__(self, bot):
-        self.bot = bot
+    def __init__(self, bot: AbstractBot):
+        self.bot: AbstractBot = bot
 
     @abc.abstractmethod
-    def is_connected(self):
-        """Tell if the backend is connected or not.
-
-        :rtype: bool
-        """
+    def is_connected(self) -> bool:
+        """Tell if the backend is connected or not."""
 
     @abc.abstractmethod
-    def on_irc_error(self, pretrigger):
+    def on_irc_error(self, pretrigger: PreTrigger) -> None:
         """Action to perform when the server sends an error event.
 
         :param pretrigger: PreTrigger object with the error event
-        :type pretrigger: :class:`sopel.trigger.PreTrigger`
 
         On IRC error, if ``bot.hasquit`` is set, the backend should close the
         connection so the bot can quit or reconnect as required.
         """
 
     @abc.abstractmethod
-    def irc_send(self, data):
+    def irc_send(self, data: bytes) -> None:
         """Send an IRC line as raw ``data``.
 
         :param bytes data: raw line to send
+
+        This method must be thread-safe.
         """
 
-    def send_command(self, *args, **kwargs):
+    @abc.abstractmethod
+    def run_forever(self) -> None:
+        """Run the backend forever (blocking call).
+
+        This method is responsible for initiating the connection to the server,
+        and it must call ``bot.on_connect`` once connected, or ``bot.on_close``
+        if it fails to connect.
+
+        Then it must run forever, listening to the server, and allowing the
+        bot to use :meth:`~.send_command` in a thread-safe way.
+        """
+
+    def decode_line(self, line: bytes) -> str:
+        """Decode a raw IRC line from ``bytes`` to ``str``."""
+        # We can't trust clients to pass valid Unicode.
+        try:
+            data = str(line, encoding='utf-8')
+        except UnicodeDecodeError:
+            # not Unicode; let's try CP-1252
+            try:
+                data = str(line, encoding='cp1252')
+            except UnicodeDecodeError:
+                # Okay, let's try ISO 8859-1
+                try:
+                    data = str(line, encoding='iso8859-1')
+                except UnicodeDecodeError:
+                    raise RuntimeError('Unable to decode data from server.')
```

Comment:
Because it doesn't make a lot of sense to log that here anyway, given that the caller of `decode_line` should be the one responsible for that: it has the raw line in bytes and an exception to deal with.

However, I don't think I log anything when calling `decode_line` later on, so that should be considered.

Reply:
Probably want to use something other than `RuntimeError`? I don't think I was clear about _what part_ of the approach I don't like, which is the lack of logging and the choice of exception type. I don't care if the caller is responsible for logging the failure, but I think something like `ValueError` would make more sense.